File: Binary Search Tree.py
# This is Binary Search Tree where left child is smaller than right
# and NOT Binary Tree.

class BinarySearchTree:
# ----------------------------------------
    class Node:
        def __init__(self, value):
            self.value = value
            self.leftChild = None
            self.rightChild = None
# ----------------------------------------
    def __init__(self):
        self.root = None
        self.inOrderTraverse = []
        self.preOrderTraverse = []
        self.postOrderTraverse = []
# ----------------------------------------
    def insert(self, value):
        node = self.Node(value)

        if self.root == None:
            self.root = node
            return
        
        current  = self.root
        while True:
            if value < current.value:
                if current.leftChild is None:
                    current.leftChild = node
                    break
                current = current.leftChild
            else:
                if current.rightChild is None:
                    current.rightChild = node
                    break
                current = current.rightChild
# ----------------------------------------
    def remove(self, value):
        self.__remove(self.root, value)
        print(f"Deleted {value}")
# ----------------------------------------
    def __remove(self, root, value):
        if root is None:
            return root
        
        if value < root.value:
            root.leftChild = self.__remove(root.leftChild, value)
        elif value > root.value:
            root.rightChild = self.__remove(root.rightChild, value)
        else:
            if root.leftChild is None:
                rightChild = root.rightChild
                root = None
                return rightChild
            elif root.rightChild is None:
                leftChild = root.leftChild
                root = None
                return leftChild
            
            rightChildLeaf = self.minValueNode(root.rightChild)
            root.value = rightChildLeaf.value
            root.rightChild = self.__remove(root.rightChild, rightChildLeaf.value)
        return root
# ----------------------------------------
    def minValueNode(self, root):
        current = root
        while(current.leftChild is not None):
            current = current.leftChild
        return current
# ----------------------------------------
    def find(self, value):
        current = self.root
        while current is not None:
            if value < current.value:
                current = current.leftChild
            elif value > current.value:
                current = current.rightChild
            else:
                return True
        return False
# ----------------------------------------
    def traversePreOrder(self):
        self.__traversePreOrder(self.root)
        print("Pre Order : ", self.preOrderTraverse)
        self.preOrderTraverse = []
# ----------------------------------------
    def __traversePreOrder(self, root):
        if root is None:
            return
        self.preOrderTraverse.append(root.value)
        self.__traversePreOrder(root.leftChild)
        self.__traversePreOrder(root.rightChild)
# ----------------------------------------
    def traverseInOrder(self):
        self.__traverseInOrder(self.root)
        print("In Order : ", self.inOrderTraverse)
        self.inOrderTraverse = []
# ----------------------------------------
    def __traverseInOrder(self, root):
        if root is None:
            return
        self.__traverseInOrder(root.leftChild)
        self.inOrderTraverse.append(root.value)

        self.__traverseInOrder(root.rightChild)
# ----------------------------------------
    def traversePostOrder(self):
        self.__traversePostOrder(self.root)
        print("Post Order : ", self.postOrderTraverse)
        self.postOrderTraverse = []
# ----------------------------------------
    def __traversePostOrder(self, root):
        if root is None:
            return
        self.__traversePostOrder(root.leftChild)
        self.__traversePostOrder(root.rightChild)
        self.postOrderTraverse.append(root.value)
# ----------------------------------------
    def traverseLevelOrder(self):
        for i in range(self.height()):
            for value in self.nodesAtDistance(i):
                print(value)
# ----------------------------------------
    def height(self):
        if self.root == None:
            return -1
        return self.__height(self.root)
# ----------------------------------------
    def __height(self, root):
        if self.isLeaf(root):
            return 0
        return 1 + max(self.__height(root.leftChild), self.__height(root.rightChild))
# ----------------------------------------
    def isLeaf(self, leaf):
        return leaf.leftChild is None and leaf.rightChild is None
# ----------------------------------------
    def nodesAtDistance(self, distance):
        List = []
        self.__nodesAtDistance(self.root, distance, List)
        return List
# ----------------------------------------
    def __nodesAtDistance(self, root, distance, List):
        if root is None:
            return

        if distance == 0:
            List.append(root.value)
            return

        self.__nodesAtDistance(root.leftChild, distance-1, List)
        self.__nodesAtDistance(root.rightChild, distance-1, List)

        return 

        



# numbers = [10, 22, 3, 40, 55, 7, 10, 76, 87, 56, 34, 23, 9, 66 ]
# numbers = [50, 30, 20, 40, 70, 60, 80]
numbers = [30, 20, 15, 5, 18, 25, 40, 35, 50, 45, 60]
# numbers = []

tree = BinarySearchTree()
for _ in numbers: tree.insert(_)


# tree.traverseInOrder()
# tree.traversePreOrder()
# tree.traversePostOrder()
# tree.remove(20)
# tree.remove(30)
# tree.remove(50)
# print(tree.find(55))

# print(tree.height())

tree.traverseLevelOrder()
print(tree.nodesAtDistance(2))

File: BinarySearchTree.cpp
#include <iostream>
#include <string>
#include <list>
#include <queue>

using std::cout;
using std::string;
using std::list;
using std::endl;
using std::max;
using std::queue;


// Python like print function
void print() { cout << '\n'; }
template <typename T, typename ...TAIL>
void print (const T &t, TAIL... tail)
{
    cout << t << ' ';
    print(tail...);
}


class BinarySearchTree
{
    private:

        class Node
        {
            public:

                int data;
                Node *leftChild;
                Node *rightChild;

                Node(int data){
                    this -> data = data;
                    leftChild = NULL;
                    rightChild = NULL;
                }
        };

        // Variables here
        Node *root;
        // End Variables

        void traversePreOrder(Node *root){
            if (root == nullptr)
                return;

            print(root -> data);
            traversePreOrder((*root).leftChild);
            traversePreOrder((*root).rightChild);
        }

        int height(Node *root){

            if (root == nullptr)
                return -1;

            if ( isLeaf(root) )
                return 0;

            int h = max(
                height(root ->leftChild), 
                height(root -> rightChild)) + 1;

            return h;
            
        }

        bool isLeaf(Node *node){
            return (node -> leftChild == nullptr && node -> rightChild == nullptr);
        }

        void getNodesAtDistance(Node *root, int distance, list<int> &nodesList){
            if (root == nullptr)
                return;

            if (distance == 0){
                nodesList.push_back(root -> data);
                return;
            }

            getNodesAtDistance(root -> leftChild, distance - 1, nodesList);
            getNodesAtDistance(root -> rightChild, distance - 1, nodesList);
        }

        list<int> getLeftView(Node *root)
        {
            list<int> leftviewlist;

            if (root == nullptr)
                return leftviewlist;
            
            queue<Node*> bfs_queue;
            bfs_queue.push(root);
            
            Node *node = nullptr;
            
            while( not bfs_queue.empty() ){
                int nodesAtCurrentLvl = bfs_queue.size();
                
                leftviewlist.push_back(bfs_queue.front() -> data);
                
                for (int i = 0; i < nodesAtCurrentLvl; i++){
                    node = bfs_queue.front();
                    
                    bfs_queue.pop();

                    // Important note:
                    {
                        // Do Not do the following
                            // if (node == NULL)
                            //     continue;
                        // Its better to check whether 
                        // the leftChild and rightChild 
                        // exists and then insert it, instead 
                        // of inserting null values, and then 
                        // checking for null because, then 
                        // you have to take care for a lot of extra
                        // edge cases, and the length of the 
                        // code will increase.
                    }
                    if (node -> leftChild)
                        bfs_queue.push(node -> leftChild);
               		if (node -> rightChild)
                        bfs_queue.push(node -> rightChild);
                    
                }
            }

            return leftviewlist;
        }

        int find_diameter(Node* root, int &diameter){
            if (root == nullptr)
                return 0;
            
            int left_height = find_diameter(root -> leftChild, diameter);
            int right_height = find_diameter(root -> rightChild, diameter);
            
            diameter = max(diameter, left_height + right_height);
            
            return 1 + max(left_height, right_height);
        }

        bool validateBST(Node *root, int range[]){
            if (root == nullptr)
                return true;
            
            if (root -> data >= range[0] and root -> data <= range[1]){
                int leftRange[] = {range[0], root -> data};
                int rightRange[] = {root -> data, range[1]};
            
                return true
                        and	validateBST(root -> leftChild, leftRange)
                        and validateBST(root -> rightChild, rightRange);
            }
            
            return false;
        }

    public:

        BinarySearchTree(){
            root = NULL;
        }

        void insert(int data){

            Node *node = new Node(data);

            if (root == NULL){
                root = node;
                return;
            }

            Node *current = root;

            while(true){
                
                if (data < (*current).data){

                    if ((*current).leftChild == NULL){
                        (*current).leftChild = node;
                        break;
                    }

                    current = (*current).leftChild;
                }

                else {   

                    if ((*current).rightChild == NULL){
                        (*current).rightChild = node;
                        break;
                    }

                    current = (*current).rightChild;
                
                }
            }
        }
        

        bool find(int data){
            Node *current = root;

            while (current != nullptr){
                if (data < (*current).data)
                    current = (*current).leftChild;
                else if (data > (*current).data)
                    current = (*current).rightChild;
                else
                    return true;
            }

            return false;
        }

        void traversePreOrder(){
            traversePreOrder(root);
        }

        int height(){
            return height(root);
        }

        list<int> getNodesAtDistance(int distance){
            list<int> nodesList;
            getNodesAtDistance(root, distance, nodesList);

            return nodesList;
        }

        void traverseLevelOrder(){
            for (int i = 0; i <= height(); i++){
                for (auto &&value : getNodesAtDistance(i))
                    print( value );
            }
        }

        list<int> traverseLevelOrderQueue(){

            list<int> nodesList;
            if (root == nullptr)
                return nodesList;

            queue<Node*> bfs_queue;
            bfs_queue.push(root);

            Node *node = nullptr;

            while( not bfs_queue.empty() ){
                node = bfs_queue.front();
                nodesList.push_back(node -> data);
                bfs_queue.pop();

                if (node -> leftChild)
                    bfs_queue.push(node -> leftChild);
                if (node -> rightChild)
                    bfs_queue.push(node -> rightChild);
            }

            return nodesList;
        }

        
        list<int> getLeftView(){
            return getLeftView(root);
        }

        

    int diameterOfBinaryTree()
    {
        int diameter;
        find_diameter(root, diameter);
        return diameter;
    }


    bool validateBST() {
        int range[] = {INT_MIN, INT_MAX};
        return true and validateBST(root, range);
    }
};



int main(int argc, char const *argv[])
{

    BinarySearchTree tree;
    tree.insert(7);
    tree.insert(4);
    tree.insert(9);
    tree.insert(1);
    tree.insert(6);
    tree.insert(8);
    tree.insert(10);
    tree.insert(1);

    // print(tree.find(10));
    print("height ->", tree.height());
    // tree.traversePreOrder();

    // for (int ele : tree.getNodesAtDistance(2)){
    //     print(ele);
    // }
    
    // tree.traverseLevelOrder();

    // for (int ele : tree.getLeftView())
    //     print(ele);

    print(tree.validateBST());

    return 0;
}
